### PR TITLE
mgr: improve reporting on unloadable modules

### PR DESCRIPTION
**Pull request by jcsp**
Fixes: http://tracker.ceph.com/issues/23358

### CODE REVIEW

File: src/mon/MgrMap.h [hunk 1]
```diff
@@ -179,6 +179,18 @@ class MgrMap
...
+  bool can_run_module(const std::string &module_name, std::string *error) const
+  {
+    for (const auto &i : available_modules) {
+      if (i.name == module_name) {
+        *error = i.error_string;
+        return i.can_run;
+      }
+    }
+
+    return false;
```
Comment (tchaikov) [inline src/mon/MgrMap.h]: shall throw a `std::logic_error` here? as `module` should be in `available_modules`

Comment (jcsp) [reply src/mon/MgrMap.h]: reasonable idea, done.

**Comment by tchaikov** [comment]
it conflicts with another PR in my test batch, dropping...

**Comment by liewegas** [comment]
<pre>
2018-04-12T21:44:11.200 INFO:tasks.workunit.client.0.smithi084.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:49: expect_false:  ceph mgr module enable foodne
2018-04-12T21:44:11.493 INFO:tasks.workunit.client.0.smithi084.stderr:Error ENOENT: all mgr daemons do not support module 'foodne', pass --force to force enablement
2018-04-12T21:44:11.509 INFO:tasks.workunit.client.0.smithi084.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:49: expect_false:  return 0
2018-04-12T21:44:11.509 INFO:tasks.workunit.client.0.smithi084.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:750: test_mon_misc:  ceph mgr module enable foodne --force
2018-04-12T21:44:11.768 INFO:tasks.ceph.mon.a.smithi084.stderr:terminate called after throwing an instance of 'std::logic_error'
2018-04-12T21:44:11.768 INFO:tasks.ceph.mon.a.smithi084.stderr:  what():  Module 'foodne' does not exist
2018-04-12T21:44:11.768 INFO:tasks.ceph.mon.a.smithi084.stderr:*** Caught signal (Aborted) **
2018-04-12T21:44:11.768 INFO:tasks.ceph.mon.a.smithi084.stderr: in thread 7f2d8102d700 thread_name:ms_dispatch
2018-04-12T21:44:11.772 INFO:tasks.ceph.mon.a.smithi084.stderr: ceph version 13.0.2-1085-gc3bb5ea (c3bb5ea91927e9d609fe6de6657c3781dca43538) mimic (dev)
2018-04-12T21:44:11.772 INFO:tasks.ceph.mon.a.smithi084.stderr: 1: (()+0x4aaf20) [0x5639ec96ef20]
2018-04-12T21:44:11.772 INFO:tasks.ceph.mon.a.smithi084.stderr: 2: (()+0x11390) [0x7f2d8be53390]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 3: (gsignal()+0x38) [0x7f2d8b0d5428]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 4: (abort()+0x16a) [0x7f2d8b0d702a]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 5: (__gnu_cxx::__verbose_terminate_handler()+0x135) [0x7f2d8ca2f0f5]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 6: (__cxxabiv1::__terminate(void (*)())+0x6) [0x7f2d8c997676]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 7: (()+0x7306c1) [0x7f2d8c9976c1]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 8: (()+0x72fbf4) [0x7f2d8c996bf4]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 9: (MgrMap::can_run_module(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*) const+0x139) [0x5639ec955239]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 10: (MgrMonitor::prepare_command(boost::intrusive_ptr<MonOpRequest>)+0xee5) [0x5639ec950b05]
2018-04-12T21:44:11.773 INFO:tasks.ceph.mon.a.smithi084.stderr: 11: (MgrMonitor::prepare_update(boost::intrusive_ptr<MonOpRequest>)+0x153) [0x5639ec951253]
2018-04-12T21:44:11.774 INFO:tasks.ceph.mon.a.smithi084.stderr: 12: (PaxosService::dispatch(boost::intrusive_ptr<MonOpRequest>)+0x916) [0x5639ec891d06]
2018-04-12T21:44:11.774 INFO:tasks.ceph.mon.a.smithi084.stderr: 13: (Monitor::handle_command(boost::intrusive_ptr<MonOpRequest>)+0x227e) [0x5639ec75f4fe]
2018-04-12T21:44:11.774 INFO:tasks.ceph.mon.a.smithi084.stderr: 14: (Monitor::dispatch_op(boost::intrusive_ptr<MonOpRequest>)+0x37d) [0x5639ec763ddd]
2018-04-12T21:44:11.774 INFO:tasks.ceph.mon.a.smithi084.stderr: 15: (Monitor::_ms_dispatch(Message*)+0x4b3) [0x5639ec7650f3]
</pre>
/a/sage-2018-04-12_21:04:24-rados-wip-sage3-testing-2018-04-12-1150-distro-basic-smithi/2390766

**Comment by tchaikov** [comment]
@jcsp ping?

**Comment by jcsp** [comment]
Pushed a fix

**Comment by jcsp** [comment]
retest this please (jenkins)

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-04-20_05:42:59-rados-wip-kefu-testing-2018-04-20-1141-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-04-20_14:33:42-rados-wip-kefu-testing-2018-04-20-1141-distro-basic-mira/

the failures are addressed by 

- #21549
- #21579

File: src/mon/MgrMonitor.cc [hunk 2]
```diff
@@ -887,6 +887,16 @@ bool MgrMonitor::prepare_command(MonOpRequestRef op)
       r = -ENOENT;
       goto out;
     }
+
+    std::string can_run_error;
+    if (force != "--force" && !pending_map.can_run_module(module, &can_run_error)) {
```
Comment (tchaikov) [inline src/mon/MgrMonitor.cc]: http://pulpito.ceph.com/kchai-2018-04-23_07:38:14-rados-wip-kefu-testing-2018-04-23-1357-distro-basic-smithi/2430354/

Comment (tchaikov) [reply src/mon/MgrMonitor.cc]: #21594 is posted to address this issue.

**Comment by epuertat** [comment]
[Mon/Mgr: improve reporting on unloadable modules](https://trello.com/c/Ha2GYHfg/127-mon-mgr-improve-reporting-on-unloadable-modules)

**Pull request merged**